servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java

```diff
@@ -312,13 +316,35 @@ static <Resp> Publisher<Resp> validateResponseAndGetPayload(final StreamingHttpR
         validateContentType(headers, expectedContentType);
         final GrpcStatusCode grpcStatusCode = extractGrpcStatusCodeFromHeaders(headers);
         if (grpcStatusCode != null) {
+            // Drain the response messageBody to make sure concurrency controller marks the request as finished.
+            // In case the grpc-status is received in headers, we expect an empty messageBody, draining should not see
+            // any other frames. However, the messageBody won't complete until after the request stream completes too.
+            final Completable drainResponse = response.messageBody().beforeOnNext(frame -> {
+                throw new GrpcStatus(INTERNAL, null, "Violation of the protocol: received unexpected " +
+                        (frame instanceof HttpHeaders ? "Trailers" : "Data") +
+                        "frame after already receiving Trailers-Only response with grpc-status: " +
+                        grpcStatusCode.value() + '(' + grpcStatusCode + ')').asException();
+            }).ignoreElements();
```

Comment:
consider adding some more context in the error msg, eg. mentioning that it happened when trailers-only were received.